### PR TITLE
VersionChooser can now select versions

### LINKED ISSUE #111: View diff between two versions
User story:

> As a code reviewer, I want to be able to view the source code diff between any two versions of an add-on I’m reviewing.

Here's an example of looking at the diff between version 1.69.0 and 1.69.1 of LastPass. URL: https://reviewers.addons-dev.allizom.org/en-US/firefox/files/compare/93833...93882/

screenshot 2019-02-01 15 31 09

### LINKED ISSUE #321: Fetch a diff of two versions when a version is selected in the VersionChooser
In order to ship #111, we want to call the API when a version is selected in the `VersionChooser` component. This action will also update the URL.

### PR DESCRIPTION
**Pull request by willdurand**
Fixes #321 
~~⚠️ depends on #385~~

I believe this also fixes #111 🏆

---

This PR adds the logic to change versions in `VersionChooser`. It also ensures version order is correct (old < new). I did it in the `VersionChooser` component because it would add too much logic into the `Compare` component, but I can move everything in one place (I would understand that we want to keep the same logic in a unique file).

Only the last commit is relevant.

### CODE REVIEW

**Comment by willdurand** [comment]
This is working but the storybook has an issue and I don't really know how to fix it yet: the component receives router props with empty `match.params`. I tried to pass the params directly but it does not work. It's the first time we have such an issue because it's the first time we render a component that depends on the routing layer in Storybook. 😿

**Comment by willdurand** [comment]
rebased!

File: stories/VersionChooser.stories.tsx [hunk 1]
```diff
@@ -8,7 +8,14 @@ import { fakeVersionsList } from '../src/test-helpers';
 import { renderWithStoreAndRouter } from './utils';
 
 const render = ({ addonId = 124, store = configureStore() } = {}) => {
-  return renderWithStoreAndRouter(<VersionChooser addonId={addonId} />, store);
+  const match = {
```
Comment (kumar303) [inline stories/VersionChooser.stories.tsx]: Could we just dispatch a real `push` action similar to how we [did it on the frontend](https://github.com/mozilla/addons-frontend/blob/0c5b2ecb3440303bcece1919305b07cf82ba391f/tests/unit/amo/sagas/test_collections.js#L617-L620)?

I don't know if we're using a Redux-aware router yet, though.

File: src/components/VersionChooser/index.tsx [hunk 2]
```diff
@@ -73,13 +113,16 @@ export class VersionChooserBase extends React.Component<Props> {
...
+  const { addonId } = ownProps.match.params;
 
   return {
-    versionsMap,
+    versionsMap: byAddonId[parseInt(addonId, 10)],
   };
 };
 
-export default connect(mapStateToProps)(VersionChooserBase);
+export default withRouter<Partial<DefaultProps> & RouterProps>(
```
Comment (kumar303) [inline src/components/VersionChooser/index.tsx]: I would feel better if this started with `PublicProps` (which would be empty) just because we will most definitely forget to add `PublicProps` to this type in case we need to add some later.

File: src/components/VersionSelect/index.tsx [hunk 3]
```diff
@@ -18,11 +20,18 @@ class VersionSelectBase extends React.Component<PublicProps> {
     withLeftArrow: false,
   };
 
+  onChange = (event: React.FormEvent) => {
+    const { value } = event.currentTarget as HTMLSelectElement;
```
Comment (kumar303) [inline src/components/VersionSelect/index.tsx]: Huh. We used to do it in Flow like this (which doesn't work):

```js
  onChange = (event: React.FormEvent<HTMLSelectElement>) => {
    const { value } = event.currentTarget;

    this.props.onChange(value);
  };
```

File: stories/VersionChooser.stories.tsx (hunk 1 above)
Comment (willdurand) [reply stories/VersionChooser.stories.tsx]: Ah, that could be an idea but we don't have a redux-aware router and I don't know if that'll be easy (see #48).

File: src/components/VersionChooser/index.tsx (hunk 2 above)
Comment (willdurand) [reply src/components/VersionChooser/index.tsx]: Oh right!

File: src/components/VersionSelect/index.tsx (hunk 3 above)
Comment (willdurand) [reply src/components/VersionSelect/index.tsx]: Yeah, looks like react-bootstrap types are.. not easy to work with.

File: stories/VersionChooser.stories.tsx (hunk 1 above)
Comment (willdurand) [reply stories/VersionChooser.stories.tsx]: I fixed this issue by using a `Route`.

**Comment by willdurand** [comment]
This is ready for a review :)

File: src/components/VersionSelect/index.tsx (hunk 3 above)
Comment (willdurand) [reply src/components/VersionSelect/index.tsx]: I found a slightly better type.

**Comment by willdurand** [comment]
I know it looks broken but it actually works :D Circle-CI isn't configured yet.

**Comment by willdurand** [comment]
Rebased!

File: src/components/VersionSelect/index.tsx [hunk 4]
```diff
@@ -1,15 +1,20 @@
 import * as React from 'react';
 import { Col, Form } from 'react-bootstrap';
+// eslint-disable-next-line import/no-unresolved
```
Comment (kumar303) [inline src/components/VersionSelect/index.tsx]: Why is this needed?

File: stories/VersionChooser.stories.tsx [hunk 5]
```diff
@@ -8,7 +9,15 @@ import { fakeVersionsList } from '../src/test-helpers';
 import { renderWithStoreAndRouter } from './utils';
 
 const render = ({ addonId = 124, store = configureStore() } = {}) => {
-  return renderWithStoreAndRouter(<VersionChooser addonId={addonId} />, store);
+  return renderWithStoreAndRouter(
+    // We wrap the `VersionChooser` component into a `Route` because it reads
+    // data from the URL (via `match.params`).
+    <Route path="/:addonId" component={VersionChooser} />,
```
[the remark this answers is not in the document]
Comment (willdurand) [reply stories/VersionChooser.stories.tsx]: Yep, that works for me.

File: src/components/VersionChooser/index.tsx [hunk 6]
```diff
@@ -9,32 +10,85 @@ import { VersionsMap, fetchVersionsList } from '../../reducers/versions';
...
+  }) => {
+    const { history, match } = this.props;
+    const { addonId, lang } = match.params;
+
+    const oldVersionId = parseInt(baseVersionId, 10);
+    const newVersionId = parseInt(headVersionId, 10);
+
+    // We make sure old version is older than the new version when user changes
+    // any of the two versions.
+    if (oldVersionId > newVersionId) {
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/components/VersionChooser/index.tsx]: Ok, I'll revert that and push a follow-up :)

File: src/components/VersionSelect/index.tsx (hunk 4 above)
Comment (willdurand) [reply src/components/VersionSelect/index.tsx]: `FormControlProps` is required by TS and only available from `react-bootstrap/lib/FormControl`, but ESLint cannot find it that way (it does not resolve the `import` correctly apparently). I could not find a better solution and did not find any useful issue about that on the ESLint plugin repo :(

File: src/components/VersionChooser/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/components/VersionChooser/index.tsx]: I filed: https://github.com/mozilla/addons-code-manager/pull/391
